Issue #12266 - InvocationType improvements and cleanups.

sbordet opened the pull request:
This is the work for the client-side.

Now the `InvocationType` can be specified at the `HttpClientTransport` level, or at the `ClientConnectionFactory.Info` level (for the dynamic transport).

Wired the `InvocationType` also for `Response.ContentSourceListener`, so that for applications that read response content using `Content.Source` and specify an `InvocationType` to `demand(Runnable)`, the implementation will honor it.

sbordet committed:
Issue #12266 - InvocationType improvements and cleanups.

This is the work for the client-side.

Now the `InvocationType` can be specified at the `HttpClientTransport` level, or at the `ClientConnectionFactory.Info` level (for the dynamic transport).

Wired the `InvocationType` also for `Response.ContentSourceListener`, so that for applications that read response content using `Content.Source` and specify an `InvocationType` to `demand(Runnable)`, the implementation will honor it.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Renamed HttpClient.getTransport() to getHttpClientTransport().

Simplified retrieval of InvocationType, now only available from the HttpClientTransport.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed FCGI parsing: onResponseHeaders() was called multiple times in case of content not yet arrived.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reverted theh renaming of `HttpClient.getTransport()`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed race condition when notifying HTTP/2 `HeadersFrame`s.

Before, `Stream.Listener.onHeaders()` was assuming that the headers were processed synchronously.
Now, `HttpReceiverOverHTTP2` process them asynchronously, with a task that declares an invocation type.
This was causing a race between the task and the code present after the call to `onHeaders()`.

Introduced `Stream.Listener.onHeaders(Stream, HeadersFrame, Callback)` to allow asynchronous processing of `HeadersFrame`s.

Optimised `HttpReceiver.responseHeaders()` to avoid creating the `Content.Source` if there is no content.

Fixed `IteratingCallback` and `HTTP2Flusher` to use `tryLock()` in `toString()` to avoid deadlocks.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/client-invocationtype-cleanups'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed IteratingCallback tests due to changes in toString().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reverted "optimization" in `HttpReceiver.responseHeaders()`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Reverted another "optimization" in `HttpReceiver.responseHeaders()`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClientTransport.java:
```diff
@@ -64,24 +65,38 @@ public interface HttpClientTransport extends ClientConnectionFactory, HttpClient
+    void setConnectionPoolFactory(ConnectionPool.Factory factory);
+
+    /**
+     * @return the {@link InvocationType} associated with this {@code HttpClientTransport}.
```
The javadoc should explain that the invocation type is going to be used to figure out how to execute the listeners.

sbordet replied:
Done.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpReceiver.java:
```diff
@@ -740,6 +755,15 @@ public void onDataAvailable()
+        public InvocationType getInvocationType()
+        {
+            Runnable demandCallback = demandCallbackRef.get();
+            if (demandCallback == null)
```
There should definitely be a comment explaining why we return the connection's invocation type when the demand callback is null, as that's quite a special trick for which I already forgot the details of.

sbordet replied:
There is not much to say: if I don't have the demand callback, I cannot extract the invocation type from it, and therefore we fall back to the connection invocation type.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:
```diff
@@ -188,7 +188,7 @@ interface AsyncContentListener extends ContentSourceListener
         default void onContentSource(Response response, Content.Source contentSource)
         {
-            Runnable demandCallback = Invocable.from(Invocable.InvocationType.NON_BLOCKING, () -> onContentSource(response, contentSource));
+            Runnable demandCallback = Invocable.from(Invocable.getInvocationType(contentSource), () -> onContentSource(response, contentSource));
```
There should probably be a comment explaining why we use the invocation type of the content source: how this trick exactly works is easy to forget.

sbordet replied:
There is no other object we can ask the invocation type to.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpClientTransportOverHTTP.java:
```diff
@@ -26,11 +26,12 @@
 
 @ManagedObject("The HTTP/1.1 client transport")
-public class HttpClientTransportOverHTTP extends AbstractConnectorHttpClientTransport
+public class HttpClientTransportOverHTTP extends AbstractConnectorHttpClientTransport implements Invocable
```
The `HttpClientTransport` super-interface already extends from `Invocable`.

sbordet replied:
Done.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/HttpClientTransportOverHTTP2.java:
```diff
@@ -42,9 +42,10 @@
 
 @ManagedObject("The HTTP/2 client transport")
-public class HttpClientTransportOverHTTP2 extends AbstractHttpClientTransport
+public class HttpClientTransportOverHTTP2 extends AbstractHttpClientTransport implements Invocable
```
The `HttpClientTransport` super-interface already extends from `Invocable`.

sbordet replied:
Done.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/ClientHTTP2StreamEndPoint.java:
```diff
@@ -63,7 +72,7 @@ public Runnable onTimeout(TimeoutException timeout, Promise<Boolean> promise)
             return null;
         }
-        return new Invocable.ReadyTask(Invocable.InvocationType.NON_BLOCKING, () ->
+        return () ->
```
You changed the invocation type here because it's now executed immediately, right?

This behavior definitely should be documented somewhere.

sbordet replied:
Yes, it's executed immediately.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java:
```diff
@@ -253,15 +269,12 @@ public Runnable onTimeout(TimeoutException failure, Promise<Boolean> promise)
-        return new Invocable.ReadyTask(Invocable.InvocationType.NON_BLOCKING, () ->
-            promise.completeWith(exchange.getRequest().abort(failure))
-        );
+        return () -> promise.completeWith(exchange.getRequest().abort(failure));
```
Again due to `onTimeout()`'s returned runnable is executed immediately? This should be on `onTimeout()`'s javadoc.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java:
```diff
@@ -253,15 +269,12 @@ public Runnable onTimeout(TimeoutException failure, Promise<Boolean> promise)
     {
-        Promise<Boolean> promise = Promise.from(failed -> callback.succeeded(), callback::failed);
-        return new Invocable.ReadyTask(Invocable.InvocationType.NON_BLOCKING, () -> responseFailure(failure, promise));
+        return () -> responseFailure(failure, Promise.from(failed -> callback.succeeded(), callback::failed));
```
Ditto.

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Channel.java:
```diff
@@ -34,6 +35,8 @@ public interface HTTP2Channel
      */
     public interface Client
     {
+        public Runnable onHeaders(Stream stream, HeadersFrame frame, Callback callback);
```
There should be javadoc explaining how the returned runnables are to be executed as sometimes it's via the execution strategy, sometimes it's via `executeImmediately`.

It's paramount to know which is executed how to make sure the implementation returns invocable runnables where needed.

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/internal/HTTP2Flusher.java:
```diff
@@ -466,12 +466,18 @@ public void dump(Appendable out, String indent) throws IOException
+        try (AutoLock ignored = lock.tryLock())
+        {
+            String held = lock.isHeldByCurrentThread() ? "" : "?";
+            return String.format("%s[%s:windowQueue=%d,frameQueue=%d,processed/pending=%d/%d]",
```
I think the `:` should only be part of the string when the lock is held.

sbordet replied:
The problem is that then the formatting string becomes `%swindowQueue`, which is difficult to parse visually and IJ marks it with a wiggle underline because it thinks it's a misspelled "swindow". 🤷🏼‍♂️

lorban commented on jetty-core/jetty-http3/jetty-http3-client-transport/src/main/java/org/eclipse/jetty/http3/client/transport/HttpClientTransportOverHTTP3.java:
```diff
@@ -39,8 +39,9 @@
+import org.eclipse.jetty.util.thread.Invocable;
 
-public class HttpClientTransportOverHTTP3 extends AbstractHttpClientTransport implements ProtocolSession.Factory
+public class HttpClientTransportOverHTTP3 extends AbstractHttpClientTransport implements ProtocolSession.Factory, Invocable
```
The `HttpClientTransport` super-interface already extends from `Invocable`.

sbordet replied:
Done.

lorban commented on jetty-core/jetty-http3/jetty-http3-client-transport/src/main/java/org/eclipse/jetty/http3/client/transport/internal/HttpChannelOverHTTP3.java:
```diff
@@ -116,4 +120,48 @@ public String toString()
             receiver);
     }
+
+    private class Listener implements Stream.Client.Listener
```
The executions of the runnable desserve to be documented.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -601,15 +601,16 @@ public String toString()
-            boolean held = lock.isHeldByCurrentThread();
-            return String.format("%s@%x{handling=%s, handled=%s, send=%s, completed=%s, request=%s}",
+            String held = lock.isHeldByCurrentThread() ? "" : "?";
+            return String.format("%s@%x[%s:handling=%s,handled=%s,send=%s,completed=%s,request=%s]",
```
I think the `:` should be part of the `held` string.

lorban commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/VirtualThreadsTest.java:
```diff
@@ -71,4 +77,71 @@ public boolean handle(Request request, Response response, Callback callback)
+                response.write(false, null, Callback.NOOP);
+                // Wait to force the client to invoke the content
+                // callback separately from the headers callback.
+                Thread.sleep(500);
```
No `sleep` please; it's better to wait until the headers event has been called.

sbordet replied:
It is going to be really complicated.
We need to wait for the headers event, but guarantee that there is a `Content.Source.read()` that returns `null`.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -918,9 +918,10 @@ public boolean reset()
+        try (AutoLock ignored = _lock.tryLock())
         {
-            return String.format("%s@%x[%s, %b, %s]", getClass().getSimpleName(), hashCode(), _state, _aborted, _failure);
+            String held = _lock.isHeldByCurrentThread() ? "" : "?";
```
`:` should be in the `held` string.

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/client-invocationtype-cleanups'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/client-invocationtype-cleanups'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/client-invocationtype-cleanups'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed HTTP/2 serialization in HttpReceiverOverHTTP2.
Fixed reset race in HTTP2Stream.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/client-invocationtype-cleanups'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed flaky tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed handling of HTTP upgrade in CoreClientUpgradeRequest.
For HTTP/2, the response may have arrived, but the exchange failed (due to the request being failed); in this case, we want to notify an UpgradeException, not a generic Exception.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Cleanups.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/ProtocolHttpUpgrader.java:
```diff
@@ -66,15 +67,14 @@ public void upgrade(Response response, EndPoint endPoint, Callback callback)
                 Destination newDestination = httpClient.resolveDestination(newOrigin);
 
-                Map<String, Object> context = new HashMap<>();
+                Map<String, Object> context = new ConcurrentHashMap<>();
```
Why is a `ConcurrentHashMap` needed?

sbordet replied:
Added comment.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java:
```diff
@@ -116,61 +116,77 @@ protected HttpChannelOverHTTP2 getHttpChannel()
+
+            responseHeaders(exchange);
+
+            callback.succeeded();
```
There is enough code before that `succeeded()` call that I would wrap it with a try/catch block.

lorban commented on jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java:
```diff
@@ -116,61 +116,77 @@ protected HttpChannelOverHTTP2 getHttpChannel()
+            HttpFields trailers = frame.getMetaData().getHttpFields();
+            trailers.forEach(exchange.getResponse()::trailer);
+        }
+        callback.succeeded();
```
There is enough code before that `succeeded()` call that I would wrap it with a try/catch block.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:
```diff
@@ -188,7 +188,10 @@ interface AsyncContentListener extends ContentSourceListener
+            // Ask the InvocationType to the Content.Source, where applications
+            // have set it there using Content.Source.demand(Runnable).
+            Invocable.InvocationType invocationType = Invocable.getInvocationType(contentSource);
+            Runnable demandCallback = Invocable.from(invocationType, () -> onContentSource(response, contentSource));
```
I think we can improve `Invocable.from` so that for any `Runnable` passed in that is not `Invocable`, it will not wrap with a `ReadyTask` if the type is `BLOCKING`.

gregw commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpReceiver.java:
```diff
@@ -447,6 +453,12 @@ private void terminateResponse(HttpExchange exchange, Result result)
+    @Override
+    public InvocationType getInvocationType()
+    {
+        return Invocable.getInvocationType(contentSource);
```
Is it worthwhile working this out once when the contentSource is set?
Probably not... but maybe???

gregw commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpReceiver.java:
```diff
@@ -623,6 +637,12 @@ public long getLength()
+        @Override
+        public InvocationType getInvocationType()
+        {
+            return Invocable.getInvocationType(source);
```
source is final, so perhaps work this out in the constructor?

sbordet replied:
We can't make it a constant because `Content.Source` delegates the `InvocationType` to the demand callback that may change across invocations of `demand(Runnable)` -- although admittedly rare.

gregw reviewed (approved):
I've a couple of minor niggles... but I've approved anyway.
